Source/Core/S2Cell.js

```diff
@@ -0,0 +1,665 @@
+}
+
+// S2 Coordinate Conversions
+
+/**
+ * @private
+ */
+function convertCellIdToFaceSiTi(cellId) {
+  var faceIJ = convertCellIdToFaceIJ(cellId);
+  var face = faceIJ[0];
+  var i = faceIJ[1];
+  var j = faceIJ[2];
+
+  var isLeaf = S2Cell.getLevel(cellId) === 30;
+  var shouldCorrect =
+    !isLeaf && (BigInt(i) ^ (cellId >> BigInt(2))) & BigInt(1); // eslint-disable-line
+  var correction = isLeaf ? 1 : shouldCorrect ? 2 : 0;
+  var si = (i << 1) + correction;
+  var ti = (j << 1) + correction;
+  return [face, si, ti];
+}
+
+/**
+ * @private
+ */
+function convertCellIdToFaceIJ(cellId) {
+  if (S2LookupPositions.length === 0) {
+    generateLookupTable();
+  }
+
+  var face = Number(cellId >> BigInt(S2PositionBits)); // eslint-disable-line
+  var bits = face & S2SwapMask;
+  var lookupMask = (1 << S2LookupBits) - 1;
+
+  var i = 0;
+  var j = 0;
+
+  for (var k = 7; k >= 0; k--) {
+    var numberOfBits = k === 7 ? S2MaxLevel - 7 * S2LookupBits : S2LookupBits;
+    var extractMask = (1 << (2 * numberOfBits)) - 1;
+    bits +=
+      Number(
+        (cellId >> BigInt(k * 2 * S2LookupBits + 1)) & BigInt(extractMask) // eslint-disable-line
+      ) << 2;
+
+    bits = S2LookupIJ[bits];
+
+    var offset = k * S2LookupBits;
+    i += (bits >> (S2LookupBits + 2)) << offset;
+    j += ((bits >> 2) & lookupMask) << offset;
+
+    bits &= S2SwapMask | S2InvertMask;
+  }
+
+  return [face, i, j];
+}
+
+/**
+ * @private
+ */
+function convertFaceSiTitoXYZ(face, si, ti) {
+  var s = convertSiTitoST(si);
+  var t = convertSiTitoST(ti);
+
+  var u = convertSTtoUV(s);
+  var v = convertSTtoUV(t);
+  return convertFaceUVtoXYZ(face, u, v);
+}
+
+/**
+ * @private
+ */
+function convertFaceUVtoXYZ(face, u, v) {
+  switch (face) {
+    case 0:
+      return new Cartesian3(1, u, v);
+    case 1:
+      return new Cartesian3(-u, 1, v);
+    case 2:
+      return new Cartesian3(-u, -v, 1);
+    case 3:
+      return new Cartesian3(-1, -v, -u);
+    case 4:
+      return new Cartesian3(v, -1, -u);
+    default:
+      return new Cartesian3(v, u, -1);
+  }
+}
+
+/**
+ * @private
+ */
+function convertSTtoUV(s) {
+  if (s >= 0.5) return (1 / 3) * (4 * s * s - 1);
+  return (1 / 3) * (1 - 4 * (1 - s) * (1 - s));
+}
+
+/**
+ * @private
+ */
+function convertSiTitoST(si) {
+  return (1.0 / S2MaxSiTi) * si;
+}
+
+/**
+ * @private
+ */
+function convertIJLeveltoBoundUV(ij, level) {
+  var result = [[], []];
+  var cellSize = getSizeIJ(level);
+  for (var d = 0; d < 2; ++d) {
+    var ijLo = ij[d] & -cellSize;
+    var ijHi = ijLo + cellSize;
+    result[d][0] = convertSTtoUV(convertIJtoSTMin(ijLo));
+    result[d][1] = convertSTtoUV(convertIJtoSTMin(ijHi));
+  }
+  return result;
+}
+
+/**
+ * @private
+ */
+function getSizeIJ(level) {
+  return (1 << (S2MaxLevel - level)) >>> 0;
+}
+
+/**
+ * @private
+ */
+function convertIJtoSTMin(i) {
+  return (1.0 / S2LimitIJ) * i;
+}
+
+// Utility Functions
+
+/**
+ * @private
+ */
+function generateLookupCell(
```

Comment:
Might be good to document the differences of this and `HilbertOrder`, e.g. this is a recursive implementation, it's an optimization used by S2, etc.